### PR TITLE
Add exactly_one function

### PR DESCRIPTION
**Pull request by Xaeroxe**
Adds a small function I found uses for personally.  Can't be added to the core library because in the error case it allocates a vec and stores data in it.

### CODE REVIEW

**Comment by Centril** [comment]
Relevant: https://github.com/rust-lang/rust/pull/55355.

**Comment by bluss** [comment]
I'd prefer this in Centril's version. And now we know, std doesn't want this method, so we can take it.

**Comment by Centril** [comment]
(Feel free to copy the contents from my PR if you want)

**Comment by Xaeroxe** [comment]
@bluss Why do we prefer returning nothing in the case of there being more than 1?  Since both versions consume the iterator it seems to me we're just needlessly hiding valuable debugging info.

**Comment by Xaeroxe** [comment]
If we simply don't like memory allocation for the error vec then I suppose we could make this take `&mut` instead and change the `Err` type to instead be `(Self::Item, Self::Item)`.

**Comment by Centril** [comment]
In all the cases where I needed this I didn't want to get the remaining elements so the allocation shouldn't be pushed on folks imo; If you want to give back the iterator you can do: `-> Result<Self::Item, (Self::Item, Self)>` and then you can use `rest` in `Err(x, rest)` to get the remaining bits if you need that.

**Comment by bluss** [comment]
We can return an iterator or similar, we already have one for iterator + one item in itertools, PutBack, so that s a nice one to put in the error case

**Comment by Xaeroxe** [comment]
@bluss I've rewritten this to use an iterator in the error case.

**Comment by bluss** [comment]
Oh right, we need to put back two elements. Either use putbackn or a custom iterator in that case, I don't mind the latter

**Comment by Centril** [comment]
I think this turned out nicely. :)

**Comment by Xaeroxe** [comment]
@bluss how's this?

**Comment by Xaeroxe** [comment]
putbackn uses a vec internally which puts us back at the original problem, so I went for a custom iterator.

**Comment by Xaeroxe** [comment]
@bluss Just pinging you in case this got lost.  Based on your github history it seems the past few days have been very busy for you.

**Comment by Xaeroxe** [comment]
@bluss Believe this is ready to merge now.

File: src/exactly_one_err.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    first_two: (Option<I::Item>, Option<I::Item>),
+    inner: I,
+}
+
+impl<I> ExactlyOneError<I>
+where
+    I: Iterator,
+{
+    /// Creates a new `ExactlyOneErr` iterator.
+    pub fn new(first_two: (Option<I::Item>, Option<I::Item>), inner: I) -> Self {
```
Comment (bluss) [inline src/exactly_one_err.rs]: this should be pub(crate) and not public, but I'll fix it up

**Comment by bluss** [comment]
Sorry, we need to add at test for this iterator before it merges. Then I agree it's ready. Can be a simple quickcheck test.

**Comment by Xaeroxe** [comment]
@bluss is the doc comment test not sufficient?

**Comment by bluss** [comment]
@Xaeroxe I realize I had thought about tests before and accepted the doctests, so I'm sorry that I didn't remember that. I shouldn't change my mind.

But in principle, no, features should have tests. Preferably quickcheck tests! They are very good at finding bugs. I feel strongly that testing documentation is often misunderstood — we test documentation to make sure that the examples we give actually work. We should make understandable examples, and they are there for documentation, not testing.

**Comment by bluss** [comment]
So since it's annoying to change one's mind, I'll merge since I had already decided some point prior to just let it slide.

**Comment by bluss** [comment]
bors r+

Thank you!

**Comment by Xaeroxe** [comment]
@bluss well I mean you made a really good case and I agree with you, so I'm happy to add the unit tests if you want.

**Comment by bluss** [comment]
@Xaeroxe ok, that can be a new PR if you want to

**Comment by bluss** [comment]
Wheels of bors are already turning

**Comment by Xaeroxe** [comment]
@bluss I've opened #324 , thanks!

**Pull request merged**